source/opt/loop_unroller.cpp

```diff
@@ -0,0 +1,793 @@
+    if (loop_induction_variable_->result_id() == old_id) {
+      // Save a pointer to the new copy of it.
+      state_.new_phi = &inst;
+    }
+  }
+}
+
+// For all instructions in |basic_block| check if the operands used are from a
+// copied
+// instruction and if so swap out the operand for the copy of it.
+void LoopUnrollerUtilsImpl::RemapOperands(ir::BasicBlock* basic_block) {
+  for (ir::Instruction& inst : *basic_block) {
+    auto remap_operands_to_new_ids = [this](uint32_t* id) {
+      auto itr = state_.new_inst.find(*id);
+      if (itr != state_.new_inst.end()) {
+        *id = itr->second;
+      }
+    };
+
+    inst.ForEachInId(remap_operands_to_new_ids);
+  }
+}
+
+// Generate the ordered list of basic blocks in the |loop| and cache it for
+// later use.
+void LoopUnrollerUtilsImpl::ComputeLoopOrderedBlocks(ir::Loop* loop) {
+  loop_blocks_inorder_.clear();
+
+  opt::DominatorAnalysis* analysis =
+      ir_context_->GetDominatorAnalysis(&function_, *ir_context_->cfg());
+  opt::DominatorTree& tree = analysis->GetDomTree();
+
+  // Starting the loop header BasicBlock, traverse the dominator tree until we
+  // reach the merge blockand add every node we traverse to the set of blocks
+  // which we consider to be the loop.
+  auto begin_itr = tree.GetTreeNode(loop->GetHeaderBlock())->df_begin();
+  const ir::BasicBlock* merge = loop->GetMergeBlock();
+  auto func = [merge, &tree, this](DominatorTreeNode* node) {
+    if (!tree.Dominates(merge->id(), node->id())) {
+      this->loop_blocks_inorder_.push_back(node->bb_);
+      return true;
+    }
+    return false;
+  };
+
+  tree.VisitChildrenIf(func, begin_itr);
+}
+
+/*
+ * End LoopUtilsImpl.
+ */
+
+}  // namespace
+
+/*
+ *
+ *  Begin Utils.
+ *
+ * */
+
+bool LoopUtils::CanPerformUnroll(ir::Loop* loop) {
+  // The loop is expected to be in structured order.
+  if (!loop->GetHeaderBlock()->GetMergeInst()) {
+    return false;
+  }
+
+  // Find check the loop has a condition we can find and evaluate.
+  const ir::BasicBlock* condition = loop->FindConditionBlock(function_);
+  if (!condition) return false;
+
+  // Check that we can find and process the induction variable.
+  const ir::Instruction* induction = loop->FindInductionVariable(condition);
+  if (!induction || induction->opcode() != SpvOpPhi) return false;
+
+  // Check that we can find the number of loop iterations.
+  if (!loop->FindNumberOfIterations(induction, &*condition->ctail(), nullptr))
+    return false;
+
+  // Make sure the continue block is a unconditional branch to the header
+  // block.
+  const ir::Instruction& branch = *loop->GetLatchBlock()->ctail();
+  bool branching_assumtion =
+      branch.opcode() == SpvOpBranch &&
+      branch.GetSingleWordInOperand(0) == loop->GetHeaderBlock()->id();
+  if (!branching_assumtion) {
+    return false;
+  }
+
+  // Make sure the induction is the only phi instruction we have in the loop
+  // header.
+  for (const ir::Instruction& inst : *loop->GetHeaderBlock()) {
```

Comment:
This seems too restrictive.  Why not just identify the phi instruction for the detected induction variable(s)?